### PR TITLE
Packaging

### PR DESCRIPTION
**Pull request by lachrimae**

### CODE REVIEW

File: platform-packaging/app/Main.hs [hunk 1]
```diff
@@ -1,16 +1,83 @@
...
+              []
+              Nothing
+              Nothing
+    putStrLn . dockerfile
+             . toDocker
+             $ env
+
+printAST :: String -> IO ()
+printAST module' = do
+    ast <- id --fmap (show . pretty)
```
Comment (TalzelAti) [inline platform-packaging/app/Main.hs]: delete debug comment

File: platform-packaging/src/Platform/Entrypoints.hs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+module Platform.Packaging.Entrypoints where
+
+import "text" Data.Text
+import "language-python" Language.Python.Common.AST
+import "language-python" Language.Python.Common.SrcLocation
+import "language-python" Language.Python.Version3.Parser (parseModule)
+
+-- findEntrypoint should find a SpanMultiLine or a SpanCoLinear
+findEntrypoints :: String -> SrcSpan
+findEntrypoint = undefined
```
Comment (TalzelAti) [inline platform-packaging/src/Platform/Entrypoints.hs]: implement or remove

File: platform-packaging/src/Platform/Entrypoints.hs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                     then Nothing
+                     else Just (f n as)
+
+take' :: Int -> Maybe [a] -> Maybe [a]
+take' = lengthCheck take
+
+drop' :: Int -> Maybe [a] -> Maybe [a]
+drop' = lengthCheck drop
+
+generateMain :: String -> Maybe String
```
Comment (TalzelAti) [inline platform-packaging/src/Platform/Entrypoints.hs]: implement or remove

File: platform-packaging/src/Platform/Packaging.hs [hunk 4]
```diff
@@ -11,39 +11,115 @@ import "base"           Data.Typeable (Typeable)
...
 import "lens"           Control.Lens hiding (from)
+import "aeson"          Data.Aeson (ToJSON, FromJSON, toEncoding, genericToEncoding, defaultOptions)
 import "data-default"   Data.Default (Default, def)
 import "containers"     Data.Map.Strict (Map, empty, assocs)
 import "text"           Data.Text (Text)
-import "dockerfile"     Data.Docker (Docker, from, run, cmd)
+import "dockerfile"     Data.Docker
 import "platform-types" Platform.Types
 
+data OS = Ubuntu
```
Comment (TalzelAti) [inline platform-packaging/src/Platform/Packaging.hs]: shouldnt types from here, move to platform-packaging-types?

File: platform-packaging/src/Platform/Packaging/Pip.hs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+pipCmd :: (Opts a)
+       => Text
+       -> GeneralOpts
+       -> a
+       -> PipInput
+       -> IO Text
+pipCmd command gOpts opts input =
+  pip $ command : fmtOpts gOpts ++ fmtOpts opts ++ fmtInput input
+
+install :: GeneralOpts
```
Comment (TalzelAti) [inline platform-packaging/src/Platform/Packaging/Pip.hs:31]: type signatures in this file can be gathered: for example:
``` haskell
install,download :: GeneralOpts
                 -> InstallOpts
                 -> PipInput
                 -> IO Text
install = pipCmd "install"
download = pipCmd "download"
```

File: platform-packaging-types/src/Platform/Packaging/Pip/Types.hs [hunk 6]
```diff
@@ -0,0 +1,1083 @@
+{-# LANGUAGE DeriveDataTypeable #-}
```
Comment (TalzelAti) [inline platform-packaging-types/src/Platform/Packaging/Pip/Types.hs:1]: split your code to:

Types/Types.hs <- holding data, type & classes declarations
Types/Default.hs <-  default instances
types/JSON.hs <- json instances
----      add as many as of the above
Types.hs <- now imports the files in the new Types folder and exposes them
Utils.hs <- an example that contains all the code from line 970

example for `Types.hs` is given here https://github.com/Atidot/platform/blob/master/platform-deployment/src/Atidot/Platform/Deployment/Interpreter/AMI/Types.hs

also, look at `Types/Types.hs` 

also see my comment about gathering type signatures in another file below, maybe it applies to the code here as well

File: platform-packaging/src/Platform/Entrypoints.hs [hunk 7]
```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (TalzelAti) [inline platform-packaging/src/Platform/Entrypoints.hs:1]: what is the purpose of this file?

File: platform-packaging/src/Platform/Packaging.hs (hunk 4 above)
Comment (lachrimae) [reply platform-packaging/src/Platform/Packaging.hs]: I've now moved the types.

File: platform-packaging/src/Platform/Entrypoints.hs (hunk 3 above)
Comment (lachrimae) [reply platform-packaging/src/Platform/Entrypoints.hs]: I removed this

File: platform-packaging/src/Platform/Entrypoints.hs (hunk 2 above)
Comment (lachrimae) [reply platform-packaging/src/Platform/Entrypoints.hs]: I removed this

File: platform-packaging/app/Main.hs (hunk 1 above)
Comment (lachrimae) [reply platform-packaging/app/Main.hs]: Debug comment deleted

**Review by TalzelAti — APPROVED** [review]

**Pull request merged**